Lib/concurrent/futures/_base.py

```diff
@@ -606,6 +627,12 @@ def result_iterator():
                 # reverse to keep finishing order
                 fs.reverse()
                 while fs:
+                    if (
+                        buffersize
+                        and (executor := executor_weakref())
+                        and (args := next(zipped_iterables, None))
+                    ):
+                        fs.appendleft(executor.submit(fn, *args))
```

Comment:
An implementation question: should we always keep a full buffer, or should we wait for the entire buffer to be full again, then we yield all buffered values?

Because now that I think about it, the buffering part is essentially useful at the very beginning to avoid iterating over the entire zip() right?

Reply:
My reasoning is as follows: During a call to `next` on the result iterator returned by `map`, we pop the oldest future from the buffer, wait for it, and yield its result. Let’s enqueue the next future immediately beforehand (keeping the buffer full) because:
- _In the best-case scenario_: all the futures already in the buffer are completed, and there is a worker in the pool available to start processing this new task, thereby reducing idle time.
- _In the worst-case scenario_: it becomes just another future pending to be processed as soon as possible, and we still respect the `buffersize` the client is comfortable with.

Do you see any scenario where this would not be the optimal behavior under the `buffersize` constraint? (I struggle to see a scenario where the "completely empty, then re-fill" behavior would not just lead to more idle and spikes 🤔 )

Reply:
I can't imagine one scenario but this was a question I wanted to ask.